src/main/java/com/microsoft/graph/content/MSBatchRequestContent.java

```diff
@@ -1,112 +1,108 @@
+import org.json.simple.parser.ParseException;
+
+import com.microsoft.graph.httpcore.RequestSerializer;
+
+import okhttp3.Headers;
+import okhttp3.RequestBody;
 
 public class MSBatchRequestContent {
-	private List<MSBatchRequestStep> batchRequestStepsArray;
-	private final int maxNumberOfRequests = 20;
+	private Map<String, MSBatchRequestStep> batchRequestStepsHashMap;
+	public static final int MAX_NUMBER_OF_REQUESTS = 20;
 	
 	public MSBatchRequestContent(List<MSBatchRequestStep> batchRequestStepsArray) {
-		this.batchRequestStepsArray = new ArrayList<>();
-		if(batchRequestStepsArray.size() <= maxNumberOfRequests) {
-			for(MSBatchRequestStep requestStep: batchRequestStepsArray)
-				addBatchRequestStep(requestStep);
-		}
+		if(batchRequestStepsArray.size() > MAX_NUMBER_OF_REQUESTS)
+			throw new IllegalArgumentException("Number of batch request steps cannot exceed " + MAX_NUMBER_OF_REQUESTS);
+		
+		this.batchRequestStepsHashMap = new HashMap<>();
+		for(MSBatchRequestStep requestStep: batchRequestStepsArray)
+			addBatchRequestStep(requestStep);
 	}
 	
 	public MSBatchRequestContent() {
-		batchRequestStepsArray = new ArrayList<>();
+		batchRequestStepsHashMap = new HashMap<String, MSBatchRequestStep>();
 	}
 	
 	public boolean addBatchRequestStep(MSBatchRequestStep batchRequestStep) {
-		if(batchRequestStep.getRequestId().compareTo("") == 0)
-			return false;
-		if(batchRequestStepsArray.size() == maxNumberOfRequests)
+		if(batchRequestStepsHashMap.containsKey(batchRequestStep.getRequestId())) 
 			return false;
-		for(MSBatchRequestStep requestStep: batchRequestStepsArray) {
-			if(batchRequestStep.getRequestId().compareTo(requestStep.getRequestId()) == 0)
-				return false;
-		}
-		return batchRequestStepsArray.add(batchRequestStep);
+		batchRequestStepsHashMap.put(batchRequestStep.getRequestId(), batchRequestStep);
+		return true;
 	}
 	
-	public boolean removeBatchRequesStepWithId(String requestId) {
-		boolean ret = false;
-		for (int i = batchRequestStepsArray.size()-1; i >= 0; i--)
-	    {
-	        MSBatchRequestStep requestStep = batchRequestStepsArray.get(i);
-	        for (int j = requestStep.getArrayOfDependsOnIds().size() - 1; j >= 0; j--)
-	        {
-	            String dependsOnId = requestStep.getArrayOfDependsOnIds().get(j);
-	            if(dependsOnId.compareTo(requestId) == 0)
-	            {
-	                requestStep.getArrayOfDependsOnIds().remove(j);
-	                ret = true;
-	            }
-	        }
-	        if(requestId.compareTo(requestStep.getRequestId()) == 0) {
-	            batchRequestStepsArray.remove(i);
-	            ret = true;
-	        }
-	    }
-		return ret;
+	public boolean removeBatchRequestStepWithId(String requestId) {
+		boolean removed = false;
+		if(batchRequestStepsHashMap.containsKey(requestId)) {
+			batchRequestStepsHashMap.remove(requestId);
+			removed = true;
+		}
+		for(Map.Entry<String, MSBatchRequestStep> steps : batchRequestStepsHashMap.entrySet()) {
```

Comment:
move it within containsKey block

Reply:
Moved loop for removing dependency in dependsOn array of each request inside containKey block.